The "Medicine" skill has gameplay effects.

afterthought2 opened the pull request:
:cl: afterthought and chinsky
rscadd: The "Medicine" skill now influences interaction with some medical objects: syringes, defibrillators, sleepers, handheld scanners, full body scanners, IV drips, and splints.
tweak: Full body scanners now store scan data. Use the menu options to print, erase, or scan/re-scan.
tweak: IV drips now take a short time to hook up.
/:cl:

Link to document with fuller description [here](https://docs.google.com/document/d/1TOnYVtz60z-rDyCISlsoErqbkWVyoq3smkcUtFOVCRw/edit?usp=sharing).

Would need wiki updates after merging. I'm not entirely happy with the scanners yet, and am generally open to suggestions. More skills coming in the future, as I get time to work on it.

afterthought2 committed:
medical changes

fix the tags

changes to unskilled and full body scanner

code improvements

do_skilled

probabilities improved

Anton-Kr commented on code/game/machinery/adv_med.dm:
```diff
@@ -239,7 +239,7 @@
 	else
 		if (src.connected) //Is something connected?
-			dat = connected.occupant.get_medical_data()
+			dat = connected.occupant.get_medical_data(user.skillset.from_path(/decl/hierarchy/skill/medical/medical))
```
Doesn't really make much sense for a computer to print something different based on a user. Would be more "sane" if the contents varied on the paper by the readers skill level.

Anton-Kr replied:
After discussion we also thought that perhaps having a "scan" button would make things a bit more clear in this situation

comma replied:
make it print a special item that shows different things depending on user, instead of just a paper.
like an xray pic or something but in spess

ghost commented:
What about the antags? Are their jobs going to restrict them from medical stuff now?

afterthought2 commented:
Antags (including on-station) have "Experienced" in all skills by default. This makes them pass all of the implemented skill checks with flying colors. If you notice this to not be the case, please report it as a bug.

hey-it-stev commented:
Not keen on time taken scaling by skill level. Would mean that doctors are strictly better at first response than actual first responders.

afterthought2 commented:
The medical skill is as much for first responders as for doctors (the anatomy skill is more doctor-specific and affects surgery). A effective corpsman would put several points into medical and have enough left over for other things, while a doctor must put points into anatomy, chemistry, and virology.

comma commented:
I think it'd be nicer if instead of hard blockers it'd be just 'yeah you can try but most likely you will make it worse'.  Some examples

>At skill level “untrained,” removing the IV rips it out, inserting the IV is impossible.

IMO let them, but with a chance of making a piercing wound instead most times.

>syringe:At skill level “untrained,” always work as though on harm intent. Can’t draw blood.

Same here for blood drawing, can try, but more stabbing wounds than actual use.

>defib: At skill level “untrained,” works as though on harm intent (attempts to shock, but does nothing).

Leave small chance of it working right IMO

afterthought2 commented:
I'll look into doing that, and also messing with the full scanner a bit, tomorrow.

hey-it-stev commented:
If that's the intention, I'd request you change the Medical skill description to say as much. Also request a change on the wiki with these changes, since not taking Medical skill will lock you out of being at all effective..

afterthought2 commented:
Quoting from the description, under "Experienced," for medical:
>You are a senior nurse or paramedic, or a practicing doctor.

I will look over it, and change the "Professional" description to also mention nurses/paramedics.

The wiki will require changes if this is merged, agreed.

babydoll commented:
> I think it'd be nicer if instead of hard blockers it'd be just 'yeah you can try but most likely you will make it worse

I disagree. I think if we make this too generous then we're wasting our time here. If someone can take 1 point in a skill for basically no cost and circumvent most of the checks, they may as well not be there in the first place.

comma commented:
Hard blockers create very silly IC situations, where no matter how dire need is, you are just physically unable to even attempt the action, because of crippling fear from not being trained.
I think that instead it should be made very risky and more likely to do more harm than good, but still usable as last ditch.
It's all about the severity of consequences for not having skills.

ThatOneGuy42 commented:
I agree with chinsky, but I do think that making skills actually have an impact is dope as fuck so good first steps

chaoko99 commented:
I would suggest untrained defib having a high chance to shock the user as well. For the unaware: "Clear" means that you should clear contact from the wounded so you don't get ZAPPED.

SparklySheep commented:
Modern AED's have a little voice that tells you when a shock is coming, so I don't think that's particularly necessary.

Goodfishie commented:
the excuse for in game defibrillators being not usable by untrained personnel is that they're not AED's afaik, and also they can shock *any* rhythm, despite irl there's only two shockable rhythms afaik

I do wish that defibrillators were attachable to a patient though. Instead of wearing it on your back, put it on the table and drag the paddles onto them, then click again for a shock.

PsiOmegaDelta commented on code/game/machinery/Sleeper.dm:
```diff
@@ -74,10 +74,10 @@
-		scan = replacetext(scan,"'notice'","'white'")
-		scan = replacetext(scan,"'warning'","'average'")
-		scan = replacetext(scan,"'danger'","'bad'")
+		var/scan = (user.skillset.from_path(/decl/hierarchy/skill/medical/medical) > SKILL_BASIC) ? medical_scan_results(occupant) : "<span class='white'><b>Contains: \the [occupant]</b></span>"
```
`user.skillset.from_path(/decl/hierarchy/skill/medical/medical)` is a fairly badly named proc, seeing that it doesn't tell what it actually does, which is return a skill level.
As it, it implies it returns an entire set of skills.

Mordeth221 commented:
What about 300 year old diona.

afterthought2 commented:
All right, made some changes:

- Defibrillators will either shock the patient (usually), shock the user, or work correctly if unskilled.
- IVs can be connected when uskilled, but have a chance of failing or hurting the patient.
- Syringes have a chance to draw blood even when unskilled. Be prepared to go through multiple syringes if you plan to do this.
- The full body medical scanner has been somewhat refactored. It now buffers scan data. There is a button to scan the patient; the scan is then saved (the quality of the scan depends on the user's skill). Then anyone may view, print, or erase that scan data, or re-scan.
- I changed the name of the proc Psi didn't like.
- The "medicine" skill description has been changed slightly to emphasize nurses/EMTs more.

BlueNexus commented:
What about IPCs, who can download as much information as they want off of the exonet, at any time?

comma commented:
I didn't see any special provisions in rules for IPC characters to hold any skills regardless of their role, so they get to play by same book.

afterthought2 commented:
There is a mechanism, currently not in use, for species-based buffs/debuffs to skill points. See the readme in #21262 for an explanation. If/when the species maintainers want, they (or I) can employ it, in some balanced fashion, to tweak species' skill points. However, this is not so relevant to this PR. If you want to see it used immediately, take it up with your species maintainer, comment on #21262 with a reasonable proposal, or make your own PR after reading that readme.

PurpleMartinJCK commented:
@afterthought2 this is a major gameplay change. I would suggest working on the wiki entry in advance as it will be vital for people to have as a resource.

PurpleMartinJCK commented on code/modules/reagents/reagent_containers/syringes.dm:
```diff
@@ -139,6 +139,9 @@
 			if(NOCLONE in T.mutations) //target done been et, no more blood in him
 				to_chat(user, "<span class='warning'>You are unable to locate any blood.</span>")
 				return
+			if((user.skillset.path2value(/decl/hierarchy/skill/medical/medical) < SKILL_BASIC) && prob(60))
```
All of these calls where it's doing these comparisons need to be moved up a level to some kind of handler. Something like "if(user.skill_min(SKILL_TYPE_MEDICAL, SKILL_LEVEL_BASIC) && prob(6))"

PurpleMartinJCK commented on code/game/objects/items/weapons/defib.dm:
```diff
@@ -320,13 +320,13 @@
 	//beginning to place the paddles on patient's chest to allow some time for people to move away to stop the process
 	user.visible_message("<span class='warning'>\The [user] begins to place [src] on [H]'s chest.</span>", "<span class='warning'>You begin to place [src] on [H]'s chest...</span>")
-	if(!do_after(user, 30, H))
+	if(!do_after(user, 3 SECONDS * user.skillset.delay_factor(/decl/hierarchy/skill/medical/medical), H))
```
These delays and things also need a more generic handler for them.

PurpleMartinJCK reviewed (changes requested):
This implementation isn't very expandable and is going to make things very complicated wherever it's used. Work on a more reasonable generic system.

afterthought2 commented:
See if you like this better, @PurpleMartinJCK. I will change the job defines to use SKILL_SKILLNAME instead of the full path, and update the documentation, in #21262 if/when this one is merged.

PurpleMartinJCK commented on code/game/objects/items/stacks/medical.dm:
```diff
@@ -266,10 +266,11 @@
-		if(do_after(user, 50, M))
-			if(M == user && prob(75))
+		var/time = 5 SECONDS * user.skill_delay_mult(SKILL_MEDICAL)
+		if(do_after(user, time, M))
```
I'd turn this into a standard thing.
"if(do_skilled(user, base_time, M, SKILL_MEDICAL))"
where do_skilled checks the skill of user for it

PurpleMartinJCK commented on code/modules/mob/skills/skill.dm:
```diff
@@ -14,6 +16,30 @@ var/const/SKILL_EASY = 1
 var/const/SKILL_AVERAGE = 2
 var/const/SKILL_HARD = 4
 
+var/const/SKILL_MANAGEMENT = /decl/hierarchy/skill/organizational/management
```
These are constants. Did it not work as
"#define SKILL_MANAGEMENT /decl/.../management"

afterthought2 replied:
Keeping them as constants for saner scoping rules and better centralization, unless you strongly object. They could be made into defines, but would have to go into a different location in the code then.

PurpleMartinJCK commented on code/game/machinery/adv_med.dm:
```diff
@@ -275,9 +247,31 @@
 		. = "moderate"
 
-/mob/living/carbon/human/proc/get_medical_data()
+/obj/machinery/body_scanconsole/proc/generate_window(mob/user)
```
Have you checked this works with the debug scanning function?

afterthought2 replied:
You mean the Analyze Health verb? It works fine with that, I checked. EDIT: now always uses max skill to generate readout.

PurpleMartinJCK reviewed (commented):
Looking way better! A couple more bits and pieces.

comma commented on code/modules/mob/skills/skillset.dm:
```diff
@@ -44,6 +38,41 @@
+	return do_after(src, base_delay * skill_delay_mult(skill_path, factor), target)
+
+// A generic way of modifying success probabilities via skill values. Higher factor means skills have more effect. fail_chance is the chance at SKILL_NONE.
+/mob/proc/skill_fail_chance(skill_path, fail_chance, factor = 1) 
```
I think it could use another argument, solid_skill or whatever, defaulting to MAX_SKILL.
It would be a skill level at which failure just doesn't happen, to prevent silly situations that non-zero chance and RNG would generate.

comma commented on code/game/objects/structures/iv_drip.dm:
```diff
@@ -169,3 +167,27 @@
+
+	if(!user.skill_check(SKILL_MEDICAL, SKILL_BASIC))
+		if(prob(60))
+			to_chat(user, "<span class='notice'>You fail to hook up \the [target] to \the [src]. Looks like you need more practice.</span>")
```
IMO only stabby effect should stay, since you're always stabbing, and plain failures without side effects are kinda eh IMO.

comma commented on code/game/objects/structures/iv_drip.dm:
```diff
@@ -169,3 +167,27 @@
+			to_chat(user, "<span class='notice'>You fail to hook up \the [target] to \the [src]. Looks like you need more practice.</span>")
+			return
+		if(prob(50))
+			visible_message("\The [user] fails to hook up \the [target] to \the [src], stabbing them instead!")
```
'fails to find the vein' so it's more explanatory on how exactly you can fail an IV

comma committed:
Tweaks to effects of medicine skill

Syringe/IV wounds now are actually piercing instead of bruising.
Failing to take blood sample doesn't break syringe. Moved it a bit so failure message is shown after start message.
Splinting others always succeeds at Trained and up

Medical handheld and advanced scanners further denied to plebs without skills.
Those without at least Basic suffer following:
-brain activity is shown as none/exists rather than degree
-oxygen/toxic damage is not shown
-tendons/arteries/fractures not shown
-internal organs aren't scanned in advanced scanner

comma commented:
https://github.com/afterthought2/Baystation12/pull/1
Got some tweaks that I thought would be faster to code than explain, pls review

babydoll commented:
I don't really feel all that comfortable reviewing this since I'm not very familiar with medical stuff at all (and I want a more restrictive thing in general), so pinging @PurpleMartinJCK since this is their speciality

HetNeSS commented:
Defibrillator successfull use requires a higher skill rate. I'm not a medic, a little bit familiar with first aid, although don't have any idea how to properly use it at all

Goodfishie commented:
if we're treating the defibs like manual defibs, I'm pretty sure basic would be fine for a defib (though afaik the main reason for using a manual is to shock when an AED can't detect a shockable pattern but the one working on the patient is certain there is one. I don't think there's much difference in terms of how they're used, but I've not used one before).
AED's (like public ones) are designed to be used by untrained people though

HetNeSS commented:
K, then. Let it be basic. But personally, i have no freaking idea where to put those paddles onto properly, as for figuring out if the revival was successful or not, as the basic wrist and neck pulse tests would, probably, not work with that matter

Goodfishie commented:
AED's tell you what to do even as someone completely untrained in them, they have a computerised voice in them that tells you what to do
they're the kind you see attached to the walls in public places irl
manual defibs are supposed to be used by people trained in them and are closer to what the defibs in the infirmary would be I'm guessing

disregarding how these defibs can shock *every* rhythm

comma commented:
it's spess tech, move on

PurpleMartinJCK reviewed (approved):
I went through and picked at a lot of things that could have skills added to them but I realised that could be done in a PR in the future. This lays a good framework and can be built up from here.
